### PR TITLE
Add `no-useless-undefined` rule

### LINKED ISSUE #683: Rule proposal: `no-useless-undefined`
# Fail

```js
function foo() {
	return undefind;
}
```

```js
let a = undefined;
```

```js
call(undefined);
```

```js
call(foo, bar, undefined);
```

```js
const {foo = undefined} = {};
```

```js
function foo({bar = undefined} = {}) {}
```

```js
function foo(bar = undefined) {}
```

# Pass

```js
function foo() {
	return;
}
```

```js
let a;
```

```js
call();
```

```js
call(foo, bar);
call(foo, bar,);
```

```js
const {foo} = {};
```

```js
function foo({bar} = {}) {}
```

```js
function foo(bar) {}
```

Discussion https://github.com/sindresorhus/eslint-plugin-unicorn/pull/636#discussion_r407150865

Not sure about this, 

```js
const noop = () => undefined;

// prefer this?
const noop = () => {};
```

### PR DESCRIPTION
**Pull request by fisker**
Fixes #683

### CODE REVIEW

**Comment by fisker** [comment]
Ready for review, hope I get everything covered, really lots of cases.

I'll fix lint issue when it's approved.

File: test/no-useless-undefined.js [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		'foo(bar,);',
+		'foo(undefined, bar);',
+		'const {foo} = {};',
+		'function foo({bar} = {}) {}',
+		'function foo(bar) {}'
+	],
+	invalid: [
+		{
+			code: 'function foo() {return undefined;}',
+			output: 'function foo() {return ;}',
```
Comment (sindresorhus) [inline test/no-useless-undefined.js:37]: I assume the space between `return` and `;` is intentional? And that the intention is that other style rules would fix it.

Comment (fisker) [reply test/no-useless-undefined.js:37]: I simply remove the `undefined` node for `return` and `yield`  cases, so the space is not taking care, do you want me to remove it?

Comment (sindresorhus) [reply test/no-useless-undefined.js:37]: Yes

Comment (fisker) [reply test/no-useless-undefined.js:37]: I tested `xo`, [`semi-spacing`](https://eslint.org/docs/rules/semi-spacing) can take care both `return ;` and `yield ;`, but I'm going to remove the space here.

Comment (fisker) [reply test/no-useless-undefined.js]: Done in https://github.com/sindresorhus/eslint-plugin-unicorn/pull/718/commits/6f4b355a1c276d7bc7007168c9653d2329ece00d

**Comment by fisker** [comment]
To make this rule work together with [`array-callback-return`](https://eslint.org/docs/rules/array-callback-return#top).

1) Disable `array-callback-return` in `recommended config`
2) Set `array-callback-return` to `[
				"error",
				{
					"allowImplicit": true
				}
			]`, but I don't think we can do this in `recommended config`, maybe in `xo` ?

[`allowImplicit`](https://eslint.org/docs/rules/array-callback-return#options)

I suggest do `2` in `xo`, WDYT?

File: test/no-useless-undefined.js [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+			output: outdent`
+				foo(
+					undefined,
+					bar,
+				)
+			`,
+			errors: [
+				{
+					messageId,
+					// The 2rd `undefined`
```
Comment (yakov116) [inline test/no-useless-undefined.js:168]: 2nd ? 3rd?

Comment (fisker) [reply test/no-useless-undefined.js:168]: 2nd, thanks

**Comment by sindresorhus** [comment]
> Set array-callback-return to [ "error", { "allowImplicit": true } ], but I don't think we can do this in recommended config, maybe in xo ?

That option is `true` by default, so XO already has that.

What we do need to do is to mention in the docs that that option should not be `false`.

**Comment by sindresorhus** [comment]
https://github.com/xojs/eslint-config-xo/commit/f3c5dd2816c3dbb98acad74cf137631dc8818297 Will be part of the next minor XO release.

**Pull request merged**